bluesky_kafka/tools/queue_thread.py

```diff
@@ -0,0 +1,213 @@
+        publisher_queue_: queue.Queue-like object
+            (name, document) tuples placed on this queue will be published as
+            Kafka messages by the publisher_
+        publisher_thread_stop_event_: threading.Event
+            the polling loop will terminate cleanly if publisher_thread_stop_event_ is set
+        publisher_queue_timeout_: float
+            time in seconds to wait for a document to become available on publisher_queue_
+            before checking if publisher_thread_stop_event_ has been set
+        """
+        name_ = None
+        document_ = None
+        published_document_count = 0
+        logger_ = logging.getLogger("bluesky_kafka")
+        logger_.info("starting Kafka message publishing loop")
+        while not publisher_thread_stop_event_.is_set():
+            try:
+                name_, document_ = publisher_queue_.get(
+                    timeout=publisher_queue_timeout_
+                )
+                publisher_(name_, document_)
+                published_document_count += 1
+            except queue.Empty:
+                # publisher_queue_.get() timed out waiting for a new document
+                # the while condition will now be checked to see if someone
+                # has requested that this thread terminate
+                # if not then try again to get a new document from publisher_queue_
+                pass
+            except BaseException:
+                # something bad happened while trying to publish a Kafka message
+                # log the exception and continue taking documents from publisher_queue_
+                logger_.exception(
+                    "an error occurred after %d successful Kafka messages when '%s' "
+                    "attempted to publish on topic %s\nname: '%s'\ndoc '%s'",
+                    published_document_count,
+                    publisher_,
+                    publisher_.topic,
+                    name_,
+                    document_,
+                )
+
+    if publisher_queue is None:
+        publisher_queue = queue.Queue()
+    publisher_thread_stop_event = threading.Event()
+    publisher_thread = threading.Thread(
+        # include a random string in the thread name in case
+        # more than one Kafka publisher thread is started
+        name=f"kafka-publisher-thread-{str(uuid.uuid4())[:8]}",
+        target=publish_documents_from_publisher_queue,
+        kwargs={
+            "publisher_": publisher,
+            "publisher_queue_": publisher_queue,
+            "publisher_thread_stop_event_": publisher_thread_stop_event,
+            "publisher_queue_timeout_": publisher_queue_timeout,
+        },
+        daemon=True,
+    )
+    publisher_thread.start()
+    logger = logging.getLogger("bluesky_kafka")
+    logger.info("Kafka publisher thread has started")
+    return publisher_queue, publisher_thread, publisher_thread_stop_event
+
+
+def build_and_start_kafka_publisher_thread(
```

Comment:
This needs a better name that says "this is the one to use to get a target for publishing".

Maybe also return `lambda name, doc: queue.add((name, doc))`?